File: ios/Classes/Data/InAppContentBlockActionCoder.swift
//
//  InAppContentBlockActionCoder.swift
//  exponea
//

import Foundation
import ExponeaSDK

class InAppContentBlockActionCoder {

        static func encode(_ source: InAppContentBlockAction) throws -> [String:Any?] {
            return [
                "type": String(describing: source.type).uppercased(),
                "name": source.name,
                "url": source.url,
            ]
        }

        static func decode(_ source: [String:Any?]) throws -> InAppContentBlockAction {
            var type: InAppContentBlockActionType = .close
            if let typeString: String = try source.getRequired("type") {
                switch typeString {
                case "DEEPLINK": type = .deeplink
                case "BROWSER": type = .browser
                case "CLOSE": type = .close
                default: throw ExponeaDataError.invalidValue(for: "type")
                }
            }
            return InAppContentBlockAction(
                name: try source.getOptional("name"),
                url: try source.getOptional("url"),
                type: type
            )
        }
    
}

File: android/src/main/kotlin/com/exponea/data/InAppContentBlockActionCoder.kt
package com.exponea.data

import com.exponea.getNullSafely
import com.exponea.getRequired
import com.exponea.sdk.models.InAppContentBlockAction
import com.exponea.sdk.models.InAppContentBlockActionType

class InAppContentBlockActionCoder {
    companion object {
        fun encode(source: InAppContentBlockAction): Map<String, Any?> {
            return mapOf(
                "type" to source.type.toString(),
                "name" to source.name,
                "url" to source.url,
            )
        }

        fun decode(source: Map<String, Any?>): InAppContentBlockAction {
            return InAppContentBlockAction(
                type = InAppContentBlockActionType.valueOf(source.getRequired("type")),
                name = source.getNullSafely("name"),
                url = source.getNullSafely("url")
            )
        }
    }
}